diff --git a/lib/ns/query.c b/lib/ns/query.c
index f5e89bd50a..3a5d5bf1ae 100644
--- a/lib/ns/query.c
+++ b/lib/ns/query.c
@@ -12176,5 +12176,11 @@ ns_query_start(ns_client_t *client, isc_nmhandle_t *handle) {
		message->flags |= DNS_MESSAGEFLAG_AD;
	}
 
+	if (qtype == dns_rdatatype_any &&
+		(client->attributes & NS_CLIENTATTR_TCP) == 0)
+	{
+		client->message->flags |= DNS_MESSAGEFLAG_TC;
+	}
+
	query_setup(client, qtype);
 }
